File: core/src/main/scala/com/graphbrain/nlp/POSTagger.scala
package com.graphbrain.nlp

import edu.stanford.nlp.tagger.maxent.MaxentTagger

class POSTagger {

	val tagger = new MaxentTagger("pos_models/english-left3words-distsim.tagger")
  //val tagger = new MaxentTagger("pos_models/french.tagger")

	def tagText(stringToTag: String): List[(String, String)] =
		taggedTokens(addTagToText(stringToTag))

	def getTokenSequence(stringToTag: String): String = {
		var tokenSeq = ""
		val tokenTags=tagText(stringToTag)
		for (tokenTag <- tokenTags) {
			tokenTag match {
				case (word, tag) => tokenSeq = tokenSeq + tag + " "
			}
		}
		tokenSeq.trim
	}

	//Returns the tagged tokens as a list of (word, tag) tuples.
	private def taggedTokens(taggedString:String):List[(String, String)]=
	{
		var taggedTokens:List[(String, String)]=List()
		val wordTagPairs = taggedString.split(" ")	
		for (wordTag <- wordTagPairs) {
			//println(wordTag);
			taggedTokens=(wordTag.split("_")(0), wordTag.split("_")(1))::taggedTokens
		}
		taggedTokens.reverse
	}

	private def addTagToText(stringToTag: String): String =
	  tagger.tagString(stringToTag)
}

object POSTagger {
	def main(args: Array[String]) = {
    val taggerTest = new POSTagger()

    val taggedToks = taggerTest.tagText("John's dog")


    for(tag <- taggedToks) {
    	tag match{
    		case (a,b) => println (a + ", " +  b)
    	}
    }
    println(taggerTest.getTokenSequence("John's dog"))
	}
}

File: src/main/java/com/graphbrain/eco/POSTagger.java
package com.graphbrain.eco;

import edu.stanford.nlp.tagger.maxent.MaxentTagger;
import java.util.List;

public class POSTagger {

    // prevent instantiation
    private POSTagger() {}

    private static MaxentTagger tagger;
    private static StanfordLemmatizer s;

    static {
        //tagger = new MaxentTagger("pos_models/english-left3words-distsim.tagger");
        tagger = new MaxentTagger("pos_models/french.tagger");
        s = new StanfordLemmatizer();
    }

    /**
     Returns a list of annotated words: (word, pos, lemma)
     */
    public static Word[] annotate(String stringToAnnotate) {
        // taggedTokens
        String taggedString = tagger.tagString(stringToAnnotate);
        String[] wordTagPairs = taggedString.split(" ");

        int length = wordTagPairs.length;
        Word[] annotated = new Word[length];

        int i = 0;
        for (String wordTag : wordTagPairs) {
            String[] tt = wordTag.split("_");
            Word word = new Word(tt[0], tt[1], "");
            annotated[i] = word;
            i++;
        }

        // lemmatise
        List<String> lemmas = s.lemmatize(stringToAnnotate, 0);

        // merge
        i = 0;
        for (String lemma : lemmas) {
            annotated[i].setLemma(lemma);
            i++;
        }

        return annotated;
    }

    public static void main(String[] args) {
        /*
        POSTagger tagger = new POSTagger();
        List<String[]> annotated = tagger.annotate("Telmo likes chocolate.");

        for (String[] elem : annotated) {
            System.out.println("-> " + elem[0] + " " + elem[1] + " " + elem[2]);
        }
        */
    }
}
